crates/templates/src/source.rs

```diff
@@ -192,6 +203,84 @@ async fn check_local(path: &Path) -> anyhow::Result<LocalTemplateSource> {
     }
 }
 
+async fn download_untar_local(url: &Url) -> anyhow::Result<LocalTemplateSource> {
+    use bytes::buf::Buf;
+
+    let temp_dir = tempdir()?;
+    let path = temp_dir.path().to_owned();
+
+    let resp = reqwest::get(url.clone())
+        .await
+        .with_context(|| format!("Failed to download from {url}"))?;
+    let tar_content = resp
+        .bytes()
+        .await
+        .with_context(|| format!("Failed to download from {url}"))?;
+
+    let reader = flate2::read::GzDecoder::new(tar_content.reader());
+    let mut archive = tar::Archive::new(reader);
+    archive
+        .unpack(&path)
+        .context("Failed to unpack tar archive")?;
+
+    let templates_root = bypass_gh_added_root(path);
+
+    Ok(LocalTemplateSource {
+        root: templates_root,
+        _temp_dir: Some(temp_dir),
+    })
+}
+
+fn bypass_gh_added_root(unpack_dir: PathBuf) -> PathBuf {
+    // GitHub adds a prefix directory to release tarballs (e.g. spin-v3.0.0/...).
+    // We try to locate the repo root within the unpacked tarball.
```

Comment:
Nit: any reason to not make this a doc comment on the function so that this surfaces more easily when code spelunking?